advisor/src/main/kotlin/advisors/Osv.kt

```diff
@@ -0,0 +1,190 @@
+import org.ossreviewtoolkit.utils.ort.log
+
+/**
+ * An advice provider that obtains vulnerability from OSV.dev https://osv.dev/.
+ */
+class Osv(name: String, advisorConfiguration: AdvisorConfiguration) : AdviceProvider(name) {
+    class Factory : AbstractAdviceProviderFactory<Osv>("Osv") {
+        override fun create(config: AdvisorConfiguration) = Osv(providerName, config)
+    }
+
+    override val details: AdvisorDetails = AdvisorDetails(providerName, enumSetOf(AdvisorCapability.VULNERABILITIES))
+
+    private val service = OsvService(
+        serverUrl = advisorConfiguration.osv?.serverUrl ?: OsvApiClient.SERVER_URL_PRODUCTION,
+        httpClient = OkHttpClientHelper.buildClient()
+    )
+
+    override suspend fun retrievePackageFindings(packages: List<Package>): Map<Package, List<AdvisorResult>> {
+        val startTime = Instant.now()
+
+        val vulnerabilitiesForPackage = getVulnerabilitiesForPackage(packages)
+
+        return packages.associateWith { pkg ->
+            val vulnerabilities = vulnerabilitiesForPackage[pkg.id].orEmpty().map { it.toOrtVulnerability() }
+
+            if (vulnerabilities.isNotEmpty()) {
+                listOf(
+                    AdvisorResult(
+                        advisor = details,
+                        summary = AdvisorSummary(
+                            startTime = startTime,
+                            endTime = Instant.now()
+                        ),
+                        vulnerabilities = vulnerabilities
+                    )
+                )
+            } else {
+                emptyList()
+            }
+        }
+    }
+
+    private fun getVulnerabilitiesForPackage(packages: List<Package>): Map<Identifier, List<Vulnerability>> {
+        val vulnerabilityIdsForPackageId = getVulnerabilityIdsForPackages(packages)
+        val allVulnerabilityIds = vulnerabilityIdsForPackageId.values.flatten().toSet()
+        val vulnerabilityForId = getVulnerabilitiesForIds(allVulnerabilityIds).associateBy { it.id }
+
+        return packages.associate { pkg ->
+            pkg.id to vulnerabilityIdsForPackageId[pkg.id].orEmpty().map { vulnerabilityForId.getValue(it) }
+        }
+    }
+
+    private fun getVulnerabilityIdsForPackages(packages: List<Package>): Map<Identifier, List<String>> {
+        val requests = packages.mapNotNull { pkg ->
+            val request = createRequest(pkg) ?: return@mapNotNull null
+            pkg to request
+        }
+
+        val result = service.getVulnerabilityIdsForPackages(requests.map { it.second })
+        if (result.isFailure) {
+            log.error {
+                "Requesting vulnerabilities IDs for packages failed: ${result.exceptionOrNull()!!.collectMessages()}"
+            }
+        }
+
+        return buildMap {
+            result.getOrNull().orEmpty().forEachIndexed { i, vulnerabilities ->
+                put(requests[i].first.id, vulnerabilities)
+            }
```

Comment:
I have a hunch that this code could be simplified by using `Result.map` which only transforms the result on success. Maybe something like `result.map { ... }.onFailure { ... }`.

Reply:
With this change, the returned map does not contain entries anymore for keys which would be associated with an empty list. There is already a discussion started about it, so let's finish this discussion first.

Reply:
I've made the change you suggested, see also discussion above.